### PR TITLE
R4R: BSC slashing

### PR DESCRIPTION
**Pull request by HaoyangLiu**

### CODE REVIEW

File: types/config.go [hunk 1]
```diff
@@ -8,5 +8,6 @@ const (
 	NativeSymbol            = "BNB"
 
 	TestnetChainID = "Binance-Chain-Nile"
+	TmpTestnetChainId = "Binance-Chain-Kongo"
```
Comment (rickyyangz) [inline types/config.go:11]: format？

File: client/rpc/staking_client.go [hunk 2]
```diff
@@ -35,6 +36,8 @@ type StakingClient interface {
 	SideChainDelegate(sideChainId string, valAddr types.ValAddress, delegation types.Coin, syncType SyncType, options ...tx.Option) (*coretypes.ResultBroadcastTx, error)
 	SideChainRedelegate(sideChainId string, valSrcAddr types.ValAddress, valDstAddr types.ValAddress, amount types.Coin, syncType SyncType, options ...tx.Option) (*coretypes.ResultBroadcastTx, error)
 	SideChainUnbond(sideChainId string, valAddr types.ValAddress, amount types.Coin, syncType SyncType, options ...tx.Option) (*coretypes.ResultBroadcastTx, error)
+	SideChainUnjail(sideChainId string, valAddr types.ValAddress, syncType SyncType, options ...tx.Option) (*coretypes.ResultBroadcastTx, error)
+	SideChainSubmitEvidence(headers [2]*bsc.Header, syncType SyncType, options ...tx.Option) (*coretypes.ResultBroadcastTx, error)
```
Comment (fletcher142) [inline client/rpc/staking_client.go:40]: better use 'bsc' as the prefix of the func name, as this is only for bsc

File: types/msg/msg-sidechain-slash.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/binance-chain/go-sdk/common/types"
+	"github.com/binance-chain/go-sdk/common/types/bsc"
+)
+
+const (
+	TypeSideChainSubmitEvidence = "submit_evidence"
```
Comment (fletcher142) [inline types/msg/msg-sidechain-slash.go:12]: --> `bsc_submit_evidence`

File: client/rpc/staking_client.go [hunk 4]
```diff
@@ -144,6 +147,29 @@ func (c *HTTP) SideChainUnbond(sideChainId string, valAddr types.ValAddress, amo
...
+	if c.key == nil {
+		return nil, KeyMissingError
+	}
+
+	m := msg.NewMsgSideChainUnjail(valAddr, sideChainId)
+
+	return c.broadcast(m, syncType, options...)
+}
+
+func (c *HTTP) SideChainSubmitEvidence(headers [2]*bsc.Header,
```
Comment (fletcher142) [inline client/rpc/staking_client.go:160]: we use `[]bsc.Header` as the type when node receiving this kind of msg

File: types/msg/msg-sidechain-stake.go [hunk 5]
```diff
@@ -13,10 +14,10 @@ const (
...
 	TypeSideChainUndelegate      = "side_undelegate"
 
-	SideChainMsgRoute    = "stake"
-	MaxSideChainIdLength = 20
-	MinSideChainAddrLen  = 16
-	MaxSideChainAddrLen  = 64
+	SideChainStakeMsgRoute = "stake"
+	MaxSideChainIdLength   = 20
+	MinSideChainAddrLen    = 16
+	MaxSideChainAddrLen    = 64
```
Comment (rickyyangz) [inline types/msg/msg-sidechain-stake.go:20]: can you help change the SideChainAddrLen check to "equals to 20"

File: go.mod [hunk 6]
```diff
@@ -5,18 +5,18 @@ require (
 	github.com/btcsuite/btcd v0.20.1-beta
 	github.com/btcsuite/btcutil v0.0.0-20190425235716-9e5f4b9a998d
 	github.com/cosmos/go-bip39 v0.0.0-20180819234021-555e2067c45d
-	github.com/go-kit/kit v0.8.0 // indirect
-	github.com/gorilla/websocket v1.4.0
+	github.com/ethereum/go-ethereum v1.9.13
```
Comment (rickyyangz) [inline go.mod:8]: do we need this dependency

File: go.sum [hunk 7]
```diff
@@ -1,8 +1,32 @@
 cloud.google.com/go v0.26.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMTw=
+github.com/Azure/azure-pipeline-go v0.2.1/go.mod h1:UGSo8XybXnIGZ3epmeBw7Jdz+HiUVpqIlpz/HKHylF4=
```
Comment (rickyyangz) [inline go.sum:2]: can we have a check about where these dependencies come from?

File: go.mod (hunk 6 above)
Comment (HaoyangLiu) [reply go.mod:8]: I provided an util to convert eth header to bsc header.

File: go.sum (hunk 7 above)
Comment (HaoyangLiu) [reply go.sum:2]:
```
github.com/ethereum/go-ethereum@v1.9.13 github.com/Azure/azure-pipeline-go@v0.2.2
```
ethereum depends on it.

File: types/msg/msg-sidechain-stake.go (hunk 5 above)
Comment (rickyyangz) [reply types/msg/msg-sidechain-stake.go:20]: I mean exactly equals to 20, instead of having min and max length check

File: common/common.go [hunk 8]
```diff
@@ -46,3 +49,32 @@ func GenerateRandomBytes(n int) ([]byte, error) {
...
+	copy(addr[:], ethHeader.Coinbase[:])
+
+	var bloom bsc.Bloom
+	copy(bloom[:], ethHeader.Bloom[:])
+
+	var blockNonce bsc.BlockNonce
+	copy(blockNonce[:], ethHeader.Nonce[:])
+
+	return bsc.Header{
+		ParentHash:  bsc.BytesToHash(ethHeader.ParentHash[:]),
```
Comment (rickyyangz) [inline common/common.go:64]: what's the difference between the original one and bytestohash one

**Pull request merged**